### ISSUE TITLE
Building libtorch in 32bit mode in Windows fails (subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--target', 'install', '--config', 'Release', '--', '/p:CL_MPCount=4']' returned non-zero exit status 1.)

### ISSUE THREAD

**Issue by Coderx7** [pytorch/pytorch#48385]
## 🐛 Bug

Trying to build libtorch in a 32bit windows 10 fails with the following error message : 
```bash
  libprotobuf-lite.vcxproj -> C:\Users\User\pytorch-1.6.0\build\lib\Release\libprotobuf-lite.lib
  Building Custom Rule C:/Users/User/pytorch-1.6.0/caffe2/CMakeLists.txt
cl : command line warning D9025: overriding '/W1' with '/w' [C:\Users\User\pytorch-1.6.0\build\caffe2\torch_global_deps.vcxproj]
  empty.c
  torch_global_deps.vcxproj -> C:\Users\User\pytorch-1.6.0\build\bin\Release\torch_global_deps.dll
Traceback (most recent call last):
  File "tools\build_libtorch.py", line 23, in <module>
    rerun_cmake=True, cmake_only=False, cmake=CMake())
  File "C:\Users\User\pytorch-1.6.0\tools\build_pytorch_libs.py", line 62, in build_caffe2
    cmake.build(my_env)
  File "C:\Users\User\pytorch-1.6.0\tools\setup_helpers\cmake.py", line 345, in build
    self.run(build_args, my_env)
  File "C:\Users\User\pytorch-1.6.0\tools\setup_helpers\cmake.py", line 141, in run
    check_call(command, cwd=self.build_dir, env=env)
  File "C:\Users\User\miniconda3\envs\env_name\lib\subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--target', 'install', '--config', 'Release', '--', '/p:CL_MPCount=4']' returned non-zero exit status 1.

```
which seems to be caused by this error that happened previously:    
```bash
 Creating library C:/Users/User/pytorch-1.6.0/build/lib/Release/torch_cpu.lib and object C:/Users/User/pytorch-1.6.0/build/lib/Release/torch_cpu.exp
  LINK : the 32-bit linker (C:\Program Files\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\bin\HostX86\x86\link.exe) ran out of heap space and is goin
  g to restart linking with a 64-bit linker
  LINK : restarting link with 64-bit linker `C:\Program Files\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\bin\HostX64\x86\link.exe'
LINK : fatal error LNK1158: cannot run 'C:\Program Files\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\bin\HostX64\x86\link.exe' [C:\Users\User\pytorc
h-1.6.0\build\caffe2\torch_cpu.vcxproj]
```

## To Reproduce

Following the instructions stated here : https://github.com/pytorch/pytorch/pull/25149 did the following: 
Inside a 32bit anaconda prompt (originally stated [here](https://github.com/planckztd/pytorch#building-the-cpu-version-libtorch-for-32-bit-on-windows) ): 
```cmd
git clone --branch v1.6.0 https://github.com/pytorch/pytorch.git pytorch-1.6.0
cd pytorch-1.6.0
git submodule update --init --recursive

:: [Optional]Only add the next line if you want to generate Libtorch's libraries in debug mode. If you want libraries in release mode, ignore the next line or set it to zero.
set DEBUG=0

:: Switch anaconda's platform from win-64 to win-32.
set CONDA_FORCE_32BIT=1

:: Recommend to create a new environment in 32-bit, and activate this environment to do all next steps.
:: Make sure in your environment, Python is in 32 bit
:: If Python shows information like this [MSC v.1912 32 bit (Intel)],which mean its architecture is 32-bit. 
conda create -n env_name python=3.7
conda activate env_name

:: Install common Dependencies.
conda install numpy ninja pyyaml mkl mkl-include setuptools cmake cffi typing

:: Disable CUDA, for just build CPU-version Libtorch
set USE_CUDA=0
set CUDAHOSTCXX=
set CMAKE_GENERATOR=Visual Studio 16 2019
set USE_NINJA=0
set USE_CUDA=0
set CMAKE_GENERATOR_TOOLSET_VERSION=
set DISTUTILS_USE_SDK=1

for /f "usebackq tokens=*" %i in (`"%ProgramFiles%\Microsoft Visual Studio\Installer\vswhere.exe" -version [15^,16^) -products * -latest -property installationPath`) do call "%i\VC\Auxiliary\Build\vcvarsall.bat" x86 -vcvars_ver=14.28

python tools\build_libtorch.py
```
## Expected behavior
Expected this to get built just fine!  

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

```
PS C:\Users\User> python .\env.py
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Enterprise
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.19.0

Python version: 3.7 (32-bit runtime)
Is CUDA available: N/A
CUDA runtime version: Could not collect
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect
```
I have the latest version of Visual Studio 2019 enterprise edition.  

Is it not possible to build this on the 32bit OS?  
Am I missing something here? 

cc @peterjc123 @maxluk @nbcsm @guyang3532 @gunandrose4u @mszhanyi @skyline75489

**Comment by skyline75489** [comment]
Based on the comments in #17865, it was possible, at least back then. However 32bit build is not supported or tested in any way.

**Comment by peterjc123** [comment]
@Coderx7 Please ensure you are using Python x86.

**Comment by Coderx7** [comment]
Hi, thanks for the reply.
I do have Python x86 installed on the system and the whole OS is 32 bit so I can only install 32 bits applications(and make sure I'm only dealing with everything x86 only)

**Comment by peterjc123** [comment]
Maybe you'll need to use the host x64 -> target x86 toolset.
```
for /f "usebackq tokens=*" %i in (`"%ProgramFiles%\Microsoft Visual Studio\Installer\vswhere.exe" -version [15^,16^) -products * -latest -property installationPath`) do call "%i\VC\Auxiliary\Build\vcvarsall.bat" amd64_x86 -vcvars_ver=14.28
```

**Comment by peterjc123** [comment]
> I do have Python x86 installed on the system and the whole OS is 32 bit so I can only install 32 bits applications(and make sure I'm only dealing with everything x86 only)

Wait, this way you could not run the cross tools. Maybe you'll need to try `MinSizeRel` builds instead.

**Comment by Coderx7** [comment]
> 
> 
> > I do have Python x86 installed on the system and the whole OS is 32 bit so I can only install 32 bits applications(and make sure I'm only dealing with everything x86 only)
> 
> Wait, this way you could not run the cross tools. Maybe you'll need to try `MinSizeRel` builds instead.

Thanks, how should I go about doing this? I was actually going to install the Win10x64 and target x86 as you just said.

**Comment by peterjc123** [comment]
@Coderx7 `set CMAKE_BUILD_TYPE=MinSizeRel`

**Comment by peterjc123** [comment]
If that doesn't work, then you may have to install a 64-bit Windows.

**Comment by Coderx7** [comment]
Thanks a lot. I tried this on a x64 bit OS and it built successfully. I followed the very same steps that mentioned in the OP with the exception that the visual studio path in the x64 bit version of Windows is a bit different (it resides in the `%%ProgramFiles(x86)%%` instead of `%%ProgramFiles%%`. i.e. I had to use this instead : 
```
for /f "usebackq tokens=*" %i in (`"%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe" -version [15^,16^) -products * -latest -property installationPath`) do call "%i\VC\Auxiliary\Build\vcvarsall.bat" x86 -vcvars_ver=14.28
```
I'll try to do this again using the 32 bit verison of Windows with the `set CMAKE_BUILD_TYPE=MinSizeRel` and see how this goes. I'll update this when its done.

**Comment by skyline75489** [comment]
To clean the cmake cache, try `python setup.py clean`. To completely clean everything, try `git clean -xdf` (this will clean every local changes, do not do this unless you really know what it does).

**Comment by Coderx7** [comment]
Ok, heres the update. using the 32 bit OS, and `CMAKE_BUILD_TYPE=MinSizeRel` doesnt work either and it will result in the same error : 
```
  LINK : the 32-bit linker (C:\Program Files\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\bin\Host
  X86\x86\link.exe) ran out of heap space and is going to restart linking with a 64-bit linker
  LINK : restarting link with 64-bit linker `C:\Program Files\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.
  28.29333\bin\HostX64\x86\link.exe'
LINK : fatal error LNK1158: cannot run 'C:\Program Files\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29
333\bin\HostX64\x86\link.exe' [C:\Users\User\pytorch-1.6.0\build\caffe2\torch_cpu.vcxproj]
```

So the only way that it seems to work is to use a 64-bit OS to build the x86.

**Comment by peterjc123** [comment]
@coderx7 Thanks for testing. I'm going to close this issue since it is resolved.

**Comment by Coderx7** [comment]
*Second update :*  
I was wrong, the commands specified here create 64bit versions. That is nothing is built as x86! 
I was trying to create the MT version (for release only as the debug seems not to be possible), but I kept getting missing libs/ conflicts. I then tried the *normal* here to see how the MD builds and what I'm doing wrong until I noticed all the generated header files are machine x64 as apposed to x86!
![image](https://user-images.githubusercontent.com/5382892/101241609-d6ee2a80-370c-11eb-8a62-84ccf992de0b.png)
